File: java/src/main/java/benchmarks/Binarytrees.java
package benchmarks;
import java.util.ArrayList;

class BinarytreesObj extends Benchmark {
    private int n;
    private long resultVal;

    static class TreeNode {
        final int item;
        final TreeNode left;
        final TreeNode right;

        TreeNode(int item, int depth) {
            this.item = item;
            if (depth > 0) {

                int shift = 1 << (depth - 1);
                left = new TreeNode(item - shift, depth - 1);
                right = new TreeNode(item + shift, depth - 1);
            } else {
                left = null;
                right = null;
            }
        }

        long sum() {
            long total = item + 1;

            if (left != null) {
                total += left.sum();
            }
            if (right != null) {
                total += right.sum();
            }

            return total;
        }
    }

    public BinarytreesObj() {
        n = (int) configVal("depth");
        resultVal = 0L;
    }

    @Override
    public String name() {
        return "Binarytrees::Obj";
    }

    @Override
    public void run(int iterationId) {
        TreeNode root = new TreeNode(0, n);
        resultVal = resultVal + root.sum();
    }

    @Override
    public long checksum() {
        return resultVal & 0xFFFFFFFFL;
    }
}

class BinarytreesArena extends Benchmark {
    private static class TreeNode {
        final int item;
        int left;
        int right;

        TreeNode(int item) {
            this.item = item;
            this.left = -1;
            this.right = -1;
        }
    }

    private ArrayList<TreeNode> arena;
    private int n;
    private long resultVal;

    public BinarytreesArena() {
        this.n = (int) configVal("depth");
        this.arena = new ArrayList<>();
        this.resultVal = 0L;
    }

    private int buildTree(int item, int depth) {
        int idx = arena.size();
        arena.add(new TreeNode(item));

        if (depth > 0) {
            int leftIdx = buildTree(item - (1 << (depth - 1)), depth - 1);
            int rightIdx = buildTree(item + (1 << (depth - 1)), depth - 1);

            TreeNode node = arena.get(idx);
            node.left = leftIdx;
            node.right = rightIdx;

        }

        return idx;
    }

    private long sum(int idx) {
        TreeNode node = arena.get(idx);
        long total = (node.item & 0xFFFFFFFFL) + 1;

        if (node.left >= 0) {
            total += sum(node.left);
        }
        if (node.right >= 0) {
            total += sum(node.right);
        }

        return total & 0xFFFFFFFFL;
    }

    @Override
    public String name() {
        return "Binarytrees::Arena";
    }

    @Override
    public void run(int iterationId) {
        arena = new ArrayList<>();
        buildTree(0, n);
        resultVal = (resultVal + sum(0)) & 0xFFFFFFFFL;
    }

    @Override
    public long checksum() {
        return resultVal & 0xFFFFFFFFL;
    }
}

File: scala/src/main/scala/benchmarks/Binarytrees.scala
package benchmarks

import scala.collection.mutable.ArrayBuffer

class BinarytreesObj extends Benchmark:
  private val n: Int = configVal("depth").toInt
  private var resultVal: Long = 0L

  override def name(): String = "Binarytrees::Obj"

  class TreeNode(val item: Int, depth: Int):
    val left: TreeNode =
      if depth > 0 then
        val shift = 1 << (depth - 1)
        TreeNode(item - shift, depth - 1)
      else null.asInstanceOf[TreeNode]

    val right: TreeNode =
      if depth > 0 then
        val shift = 1 << (depth - 1)
        TreeNode(item + shift, depth - 1)
      else null.asInstanceOf[TreeNode]

    def sum(): Long =
      var total: Long = item + 1L
      if left != null then total += left.sum()
      if right != null then total += right.sum()
      total

  override def run(iterationId: Int): Unit =
    val root = TreeNode(0, n)
    resultVal = (resultVal + root.sum()) & 0xffffffffL

  override def checksum(): Long = resultVal & 0xffffffffL

class BinarytreesArena extends Benchmark:
  private val n: Int = configVal("depth").toInt
  private var resultVal: Long = 0L

  override def name(): String = "Binarytrees::Arena"

  class TreeNodeArena(val item: Int):
    var left: Int = -1
    var right: Int = -1

  class TreeArena:
    private val nodes = ArrayBuffer.empty[TreeNodeArena]

    def build(item: Int, depth: Int): Int =
      val idx = nodes.length
      nodes += TreeNodeArena(item)

      if depth > 0 then
        val shift = 1 << (depth - 1)
        val leftIdx = build(item - shift, depth - 1)
        val rightIdx = build(item + shift, depth - 1)
        nodes(idx).left = leftIdx
        nodes(idx).right = rightIdx

      idx

    def sum(idx: Int): Long =
      val node = nodes(idx)
      var total = node.item + 1L

      if node.left >= 0 then total += sum(node.left)
      if node.right >= 0 then total += sum(node.right)

      total

  override def run(iterationId: Int): Unit =
    val arena = TreeArena()
    val rootIdx = arena.build(0, n)
    resultVal = (resultVal + arena.sum(rootIdx)) & 0xffffffffL

  override def checksum(): Long = resultVal & 0xffffffffL
